woocommerce/woocommerce
# [Settings] Create a route file to handle populating Settings

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





Closes https://github.com/woocommerce/woocommerce/issues/52276


This PR introduces a new TypeScript-based routing structure for the settings layout inspired by patterns from the [Product Editor](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/product-editor/src/products-app/router.tsx), [Site Editor](https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-site/src/components/layout/router.js) and [Proof of concept](https://github.com/woocommerce/woocommerce/blob/09cdb35a1ebbc422f87c5dea1795b0b7824d254e/plugins/woocommerce-admin/client/settings/routes.js) and adds a new `useActiveRoute` hook to determine and return the active route based on the current path.

Note: `window.wcSettings?.admin?.settingsPages` will be changed to a more useful shape in #52691 (https://github.com/woocommerce/woocommerce/issues/52276#issuecomment-2467037834). This PR assumes the new shape of `window.wcSettings?.admin?.settingsPages` is already available.



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

This PR is part of Settings project and behind a feature flag. **This doesn't need to be tested in QA testing since it's still in development and might change by follow-up PRs.**

It's challenging to test modern pages because they have not been implemented yet, but I have added tests for the `useActiveRoute` and `useModernRoutes` hooks to ensure they work as expected.

1. Create a fresh site and activate the feature flag `settings`
2. Install and activate the `Gutenberg` plugin
3. Go to `/wp-admin/admin.php?page=wc-settings`
4. Verify general setting page is displayed
5. Go to `/wp-admin/admin.php?page=wc-settings&tab=shipping` or other existing pages like ` /wp-admin/admin.php?page=wc-settings&tab=products`
6. Verify the page is displayed
7. Go to  `/wp-admin/admin.php?page=wc-settings&path=non-exist`
8. Verify Not found page is displayed.


<img width="1147" alt="Screenshot 2024-11-11 at 14 45 10" src="https://github.com/user-attachments/assets/49b38ae7-7fbf-41bd-abbb-46595b948a49">

<img width="1147" alt="Screenshot 2024-11-11 at 14 45 24" src="https://github.com/user-attachments/assets/ac9f8dbe-300c-4a9a-8b9e-9545f8581ade">



### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
